File: sdks/go/cmd/elydora/plugins/gemini_command.go
package plugins

import (
	"fmt"
	"path/filepath"
	"runtime"
	"strings"
)

type geminiRuntimeReference = managedScriptReference

func buildGeminiCommand(runtimePath, scriptPath string) (string, error) {
	if !filepath.IsAbs(runtimePath) || !filepath.IsAbs(scriptPath) {
		return "", fmt.Errorf("Gemini CLI hook commands require absolute runtime and script paths")
	}
	if runtime.GOOS != "windows" {
		return posixSource(runtimePath, scriptPath), nil
	}
	return "& " + quotePowerShellArgument(windowsPowerShellPath()) + powerShellEncodedFlags +
		encodePowerShellSource(powerShellSource(runtimePath, scriptPath)), nil
}

func parseGeminiWindowsCommand(command string) (string, string, bool) {
	if !strings.HasPrefix(command, "& ") {
		return "", "", false
	}
	powerShell, next, ok := readPowerShellArgument(command, 2)
	if !ok || !isPowerShellExecutable(powerShell) ||
		!strings.HasPrefix(command[next:], powerShellEncodedFlags) {
		return "", "", false
	}
	source, ok := decodePowerShellSource(command[next+len(powerShellEncodedFlags):])
	if !ok {
		return "", "", false
	}
	return parsePowerShellSource(source)
}

func parseGeminiCommand(command string) (string, string, bool) {
	if runtimePath, scriptPath, ok := parsePOSIXCommand(command); ok {
		return runtimePath, scriptPath, true
	}
	return parseGeminiWindowsCommand(command)
}

func parseLegacyGeminiCommand(command string) (string, string, bool) {
	if !strings.HasPrefix(command, "node ") || strings.ContainsAny(command, "\r\n") {
		return "", "", false
	}
	scriptPath := strings.TrimPrefix(command, "node ")
	if scriptPath == "" || strings.TrimSpace(scriptPath) != scriptPath {
		return "", "", false
	}
	return "node", scriptPath, true
}

func geminiRuntimeReferenceForCommand(
	command string,
	scriptName string,
	includeLegacy bool,
) (*geminiRuntimeReference, error) {
	runtimePath, scriptPath, current := parseGeminiCommand(command)
	legacy := false
	if !current && includeLegacy {
		runtimePath, scriptPath, legacy = parseLegacyGeminiCommand(command)
	}
	validRuntime := (current && filepath.IsAbs(runtimePath) && isNodeExecutable(runtimePath)) ||
		(legacy && runtimePath == "node")
	if !validRuntime {
		return nil, nil
	}
	return resolveManagedScript(scriptPath, scriptName)
}

File: sdks/node/src/plugins/gemini-command.ts
import os from 'node:os';
import path from 'node:path';
import { samePath } from './common.js';
import {
  decodePowerShellSource,
  encodePowerShellSource,
  isNodeExecutable,
  isPowerShellExecutable,
  parsePosixCommand,
  parsePowerShellSource,
  posixSource,
  quotePowerShell,
  windowsPowerShellPath,
} from './shell-command.js';

export interface GeminiRuntimeReference {
  readonly agentId: string;
  readonly scriptPath: string;
}

function windowsCommand(scriptPath: string): string {
  return `& ${quotePowerShell(windowsPowerShellPath())} -NoLogo -NoProfile -NonInteractive -EncodedCommand ${encodePowerShellSource(scriptPath)}`;
}

export function buildGeminiCommand(scriptPath: string): string {
  if (!path.isAbsolute(process.execPath) || !path.isAbsolute(scriptPath)) {
    throw new Error('Gemini CLI hook commands require absolute executable and script paths');
  }
  return process.platform === 'win32'
    ? windowsCommand(scriptPath)
    : posixSource(scriptPath);
}

function parseWindowsCommand(command: string): readonly [string, string] | undefined {
  const match = /^& ('(?:[^']|'')+') -NoLogo -NoProfile -NonInteractive -EncodedCommand ([A-Za-z0-9+/]+={0,2})$/.exec(command);
  if (!match) return undefined;
  const powershell = parsePowerShellSource(`& ${match[1]} ''; exit $LASTEXITCODE`);
  if (!powershell || !isPowerShellExecutable(powershell[0])) return undefined;
  return decodePowerShellSource(match[2]);
}

function parseLegacyCommand(command: string): readonly [string, string] | undefined {
  const match = /^node "([^"\r\n]+)"$/.exec(command);
  return match ? ['node', match[1]] : undefined;
}

export function geminiRuntimeReference(
  command: string,
  scriptName: string,
  includeLegacy = false,
): GeminiRuntimeReference | undefined {
  const parsed = parsePosixCommand(command)
    ?? parseWindowsCommand(command)
    ?? (includeLegacy ? parseLegacyCommand(command) : undefined);
  if (!parsed
    || !isNodeExecutable(parsed[0])
    || !path.isAbsolute(parsed[1])
    || path.basename(parsed[1]) !== scriptName) return undefined;
  if (parsed[0] !== 'node' && !path.isAbsolute(parsed[0])) return undefined;
  const agentDirectory = path.dirname(parsed[1]);
  if (!samePath(path.dirname(agentDirectory), path.join(os.homedir(), '.elydora'))) {
    return undefined;
  }
  const agentId = path.basename(agentDirectory);
  if (!agentId || agentId === '.' || agentId === '..') return undefined;
  return { agentId, scriptPath: parsed[1] };
}
